### PR TITLE
vtk integration for mesh processing

### PR DESCRIPTION
**Pull request by benjaminum**
This PR adds the VTK library for mesh processing and
- implements a mesh clipping operation with a plane
- adds building vtk from source as external project
- adds a workflow for precompiling vtk for linux/win/mac
- adds a build path that downloads the precompiled vtk (default)

---
This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d/4957)

### CODE REVIEW

**Comment by benjaminum** [comment]
The following probably should be discussed
- `VtkUtils.h` is placed in the `kernels` subdir and has its own namespace `vtkutils`
- arm64 has no precompiled vtk packages

File: cpp/open3d/t/geometry/kernel/VtkUtils.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#include "open3d/t/geometry/Geometry.h"
+#include "open3d/t/geometry/PointCloud.h"
+#include "open3d/t/geometry/TriangleMesh.h"
+#include "vtkPolyData.h"
```
Comment (yxlao) [inline cpp/open3d/t/geometry/kernel/VtkUtils.h:30]: Use `#include <>` for third-party headers.

File: cpp/open3d/t/geometry/kernel/VtkUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#include "open3d/t/geometry/kernel/VtkUtils.h"
+
+#include "vtkArrayDispatch.h"
```
Comment (yxlao) [inline cpp/open3d/t/geometry/kernel/VtkUtils.cpp:29]: Use `#include <>` for third-party headers.

File: cpp/open3d/t/geometry/TriangleMesh.cpp [hunk 3]
```diff
@@ -36,6 +36,10 @@
 #include "open3d/core/TensorCheck.h"
 #include "open3d/t/geometry/kernel/PointCloud.h"
 #include "open3d/t/geometry/kernel/Transform.h"
+#include "open3d/t/geometry/kernel/VtkUtils.h"
+#include "vtkCleanPolyData.h"
```
Comment (yxlao) [inline cpp/open3d/t/geometry/TriangleMesh.cpp:40]: Use `#include <>` for third-party headers.

File: cpp/open3d/t/geometry/TriangleMesh.cpp [hunk 4]
```diff
@@ -270,6 +274,35 @@ TriangleMesh TriangleMesh::To(const core::Device &device, bool copy) const {
...
+    using namespace kernel::vtkutils;
+    core::AssertTensorShape(point, {3});
+    core::AssertTensorShape(normal, {3});
+    // allow int types for convenience
+    core::AssertTensorDtypes(
+            point, {core::Float32, core::Float64, core::Int32, core::Int64});
+    core::AssertTensorDtypes(
+            normal, {core::Float32, core::Float64, core::Int32, core::Int64});
+
+    auto point_ = point.To(core::Device(), core::Dtype::Float64).Contiguous();
```
Comment (yxlao) [inline cpp/open3d/t/geometry/TriangleMesh.cpp:288]: `core::Float64`, without the `Dtype`, same for the rest.

File: cpp/open3d/t/geometry/kernel/VtkUtils.cpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    auto tensor_cpu = tensor.To(core::Device()).Contiguous();
+    const size_t num_points = tensor.GetLength();
+
+    vtkSmartPointer<vtkPoints> pts = vtkSmartPointer<vtkPoints>::New();
+
+    if (tensor.GetDtype() == core::Float32) {
+        pts->SetDataTypeToFloat();
+        pts->SetNumberOfPoints(num_points);
+        const float* data = tensor_cpu.GetDataPtr<float>();
+        for (size_t i = 0; i < num_points; ++i) {
```
Comment (yxlao) [inline cpp/open3d/t/geometry/kernel/VtkUtils.cpp:55]: Is there away to create vtk points from a buffer?

File: cpp/open3d/t/geometry/kernel/VtkUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        }
+    }
+    return pts;
+}
+
+vtkSmartPointer<vtkCellArray> CreateVtkCellArrayFromTensor(
+        const core::Tensor& tensor) {
+    vtkSmartPointer<vtkCellArray> cells = vtkSmartPointer<vtkCellArray>::New();
+
+    core::AssertTensorShape(tensor, {utility::nullopt, utility::nullopt});
```
Comment (yxlao) [inline cpp/open3d/t/geometry/kernel/VtkUtils.cpp:75]: `utility::nullopt` -> `core::None`, same in this PR.

File: cpp/open3d/t/geometry/kernel/VtkUtils.cpp [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+            data += 3;
+        }
+    } else if (points->GetDataType() == VTK_DOUBLE) {
+        result = core::Tensor({num_points, 3}, core::Float64);
+        double* data = result.GetDataPtr<double>();
+        for (int64_t i = 0; i < num_points; ++i) {
+            points->GetPoint(i, data);
+            data += 3;
+        }
+
```
Comment (yxlao) [inline cpp/open3d/t/geometry/kernel/VtkUtils.cpp:122]: nit; remove extra line

**Review by yxlao — CHANGES_REQUESTED** [review]

File: cpp/open3d/t/geometry/kernel/VtkUtils.cpp (hunk 5 above)
Comment (benjaminum) [reply cpp/open3d/t/geometry/kernel/VtkUtils.cpp:55]: I did some more research on this. It seems possible but we need to use a different array interface.
I think it is better to create a separate PR for this.

**Review by yxlao — APPROVED** [review]

**Pull request merged**